trantor/net/inner/tlsprovider/BotanTLSProvider.cc

```diff
@@ -0,0 +1,389 @@
+                contextPtr_->requireClientCert;
+            channel_ = std::make_unique<Botan::TLS::Client>(
+                *this,
+                sessionManager,
+                *credsPtr_,
+                validationPolicy_,
+                rng,
+                Botan::TLS::Server_Information(policyPtr_->getHostname(),
+                                               conn_->peerAddr().toPort()),
+                policyPtr_->getUseOldTLS()
+                    ? Botan::TLS::Protocol_Version::TLS_V10
+                    : Botan::TLS::Protocol_Version::TLS_V12,
+                policyPtr_->getAlpnProtocols());
+            setSniName(policyPtr_->getHostname());
+        }
+    }
+
+    void handleSSLError(SSLError err)
+    {
+        if (!errorCallback_)
+            return;
+        loop_->queueInLoop([this, err]() { errorCallback_(conn_, err); });
+    }
+
+    virtual ~BotanTLSProvider() override = default;
+
+    void tls_emit_data(const uint8_t data[], size_t size) override
+    {
+        auto n = writeCallback_(conn_, data, size);
+        lastWriteSize_ = n;
+
+        // store the unsent data and send it later
+        if (n == ssize_t(size))
+            return;
+        if (n == -1)
+            n = 0;
+        appendToWriteBuffer((const char *)data + n, size - n);
+    }
+
+    void tls_record_received(uint64_t seq_no,
+                             const uint8_t data[],
+                             size_t size) override
+    {
+        (void)seq_no;
+        recvBuffer_.append((const char *)data, size);
+        if (messageCallback_)
+            messageCallback_(conn_, &recvBuffer_);
+    }
+
+    void tls_alert(Botan::TLS::Alert alert) override
+    {
+        if (alert.type() == Botan::TLS::Alert::CLOSE_NOTIFY)
+        {
+            LOG_WARN << "TLS close notify received";
+            if (closeCallback_)
+                closeCallback_(conn_);
+        }
+        else
+        {
+            if (errorCallback_)
+                errorCallback_(conn_, SSLError::kSSLProtocolError);
+        }
+    }
+
+    bool tls_session_established(const Botan::TLS::Session &session)
+    {
+        (void)session;
+        LOG_TRACE << "tls_session_established";
+        tlsConnected_ = true;
+        loop_->queueInLoop([this]() {
+            setApplicationProtocol(channel_->application_protocol());
+            if (handshakeCallback_)
+                handshakeCallback_(conn_);
+        });
+        // Do we want to cache all sessions?
+        return true;
+    }
+
+    void tls_verify_cert_chain(
+        const std::vector<Botan::X509_Certificate> &certs,
+        const std::vector<std::shared_ptr<const Botan::OCSP::Response>> &ocsp,
+        const std::vector<Botan::Certificate_Store *> &trusted_roots,
+        Botan::Usage_Type usage,
+        const std::string &hostname,
+        const Botan::TLS::Policy &policy)
+    {
+        setSniName(hostname);
+        if (policyPtr_->getValidate() && !policyPtr_->getAllowBrokenChain())
```

Comment:
If `getAllowBrokenChain()` returns `true`, it seems that you will always just do the "reduced" check below ([discussed here](https://github.com/an-tao/trantor/pull/238#discussion_r1128673362) as well). Is that really what you want to do? Personally, I would have expected this policy to enable an additional fallback.

Like so maybe?

```C++
if (policyPtr_->getValidate())
{
   try
   {
      Botan::TLS::Callbacks::tls_verify_cert_chain(...)
   }
   catch (const std::exception&)
   {
      if (!policyPtr_->getAllowBrokenChain())
         throw;

      // perform the reduced check (as sketched in the other comment)
      // and throw if this fails as well
   }
}
```

Reply:
Thanks for pointing our the reduced check. That's a lot of vulnerabilities avoided. I've patched and tested the code to with our unit testes. I'll do more thorough tests in the weekends. 

I don't see how the code in your comment produced different behavior. The try catch logic reduces to the same thing as my if-else. If `getAllowBrokenChain()` is `true`. Then any exception thrown by `Botan::TLS::Callbacks::tls_verify_cert_chain` gets cough and it does the reduced check. And since the reduced check is a subset of the full check. This effectively is the same as just doing the reduced one.

Is there some global state or something I'm missing?

|             | validate | validate + allow broken | no validate |
|-------------|----------|-------------------------|-------------|
| Valid cert  | ok       | ok                      | ok          |
| self signed |          | ok                      | ok          |
| expired     |          |                         | ok          |

Reply:
I agree, the final result is the same: Broken chains pass the fallback and therefor the overall check. Fair point. I was confused by the control flow.

As a suggestion: Would it make sense to log a warning if validation is enabled and a broken chain was accepted only due to `getAllowBrokenChain()` returning `true`?

Reply:
As a side-note: We're [working on an implementation for RFC 7250](https://github.com/randombit/botan/pull/3355) to (optionally) allow authentication with raw public keys instead of certificate chains. It might take a while until this lands in a public release, but it could be a better fit for TOFU-like use cases than self-signed certs.